### PR TITLE
Imviz: Annulus background for photometry

### LINKED ISSUE #1222: Imviz: line profile interferes with radial profile
**Reporter: Camilla Pacifici**

In jdaviz 2.4.0.
I asked for a radial profile after having looked at the line profiles and the two plots showed up together.

[🐱](https://jira.stsci.edu/browse/JDAT-2291)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #1223: Imviz: radial profile plotted beyond aperture?
**Reporter: Camilla Pacifici**

Seems like the radial profiles are plotted beyond the aperture radius (points at zero).
It would be best to set the maximum xlimit of the plot to the largest radius.

[🐱](https://jira.stsci.edu/browse/JDAT-2292)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to implement annulus background calculation for Simple Aperture Photometry plugin.

| Manual | Annulus | Subset |
| --- | --- | --- |
| Screenshot 2022-03-31 170632 | Screenshot 2022-03-31 170557 | Screenshot 2022-03-31 170651 |

Also see #1050

### TODO

- [x] Rebase after #1219 is merged because this PR builds on that one.
- [x] Write tests. Make sure annulus calculation is correct for different scenarios (within reason).
- [x] Stretch goal 1 (fix, test, add change log): Fix #1222
- [x] Stretch goal 2 (fix, test, add change log): Fix #1223
- [x] Rebase again after #1221 is merged and make sure nothing breaks here.
- [x] Tag Cami and Larry for review too.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-2172) [🐱](https://jira.stsci.edu/browse/JDAT-2291) [🐱](https://jira.stsci.edu/browse/JDAT-2292)

### CODE REVIEW

**Comment by pllim** [comment]
@kecnry , do you know why setting the Data via API here is not sticking anymore? When I run this test manually, there seems to be a lag to it showing up on the GUI. Do you think this lag + async is causing the test failure? How should I fix this?

```
    def test_annulus_background(imviz_helper):
        from photutils.datasets import make_4gaussians_image
    
        gauss4 = make_4gaussians_image()  # The background has a mean of 5 with noise
        ones = np.ones(gauss4.shape)
    
        imviz_helper.load_data(gauss4, data_label='four_gaussians')
        imviz_helper.load_data(ones, data_label='ones')
    
        phot_plugin = imviz_helper.app.get_tray_item_from_name('imviz-aper-phot-simple')
        phot_plugin.data_selected = 'ones'
    
        # Mark an object of interest
        imviz_helper._apply_interactive_region('bqplot:circle', (143, 18), (157, 32))
        phot_plugin.subset_selected = 'Subset 1'
        phot_plugin.bg_subset_selected = 'Annulus'
    
        # Check annulus for ones
        assert_allclose(phot_plugin.bg_annulus_inner_r, 7)  # From circle
        assert_allclose(phot_plugin.bg_annulus_width, 10)  # Default
>       assert_allclose(phot_plugin.background_value, 1)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0
E       
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: 4.74559613
E       Max relative difference: 4.74559613
E        x: array(5.745596)
E        y: array(1)
```

**Comment by kecnry** [comment]
Assuming this was after the rebase, you need to set `phot_plugin.dataset_selected` not `phot_plugin.data_selected`.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:52]: Is there somehow we can clarify (briefly) that this means outer_radius = inner_radius + width?  Without any visual feedback, I actually had to look in the code to make sure, even though I realize it probably would have been labeled outer radius if that was what it actually meant.

File: jdaviz/utils.py
Comment (kecnry) [inline jdaviz/utils.py:89]: interesting.  Would using the object model instead of pyplot syntax get around this?

File: docs/imviz/plugins.rst
Comment (kecnry) [inline docs/imviz/plugins.rst:104]: is the display feedback out-of-scope?  Is it planned for future development?  After trying to use this for testing, I think visual feedback would be extremely useful to ensure you're doing what you think you're doing - especially with rectangular apertures and crowded fields where you want to make sure the annulus does not include other sources, etc.

**Review by kecnry — COMMENTED** [review]
Haven't managed to completely break it... but there are a lot of snackbar errors that can become quite cumbersome.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:181]: It is a general catch-all. Like if someone tries to grab annulus with all NaNs in it or whatever crazy use case that I can not yet imagine.

> a very long queue of snackbar messages

But how? 😆

File: jdaviz/utils.py
Comment (pllim) [reply jdaviz/utils.py:89]: Maybe? I am not familiar with bqplot at all, so maybe @maartenbreddels can answer. Regardless, I don't have time to do plot refactoring before the end of sprint.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:181]: I think I just changed the input multiple times that each raised an error here and they built up and took forever to clear because of the timeouts.

I think the catch-all here is probably fine, but any scenarios that we can predict, should avoid the snackbar if they are caught by validation rules (like a negative or 0 input for width).

File: docs/imviz/plugins.rst
Comment (pllim) [reply docs/imviz/plugins.rst:104]: Yes, POs specifically said do not display the annulus this round. Getting annulus to display, etc, is a lot of work because it involves new feature for glue-astronomy translator:

* glue-viz/glue-astronomy#49

Comment (pllim) [reply docs/imviz/plugins.rst:104]: Also, this plugin is not for crowded field photometry, which would probably need PSF photometry that is out of scope.

Comment (kecnry) [reply docs/imviz/plugins.rst:104]: ok, another option (either now or later) would be just lines representing the inner and outer radii of the annulus.  I suspect that would be much easier?

Comment (pllim) [reply docs/imviz/plugins.rst:104]: Re: Just lines -- I thought about that but the only API I can think of is the marker and I find the following unsolved issue too disruptive to include (in addition to all the linking and de-linking overhead that comes with adding marker table):

* #760

Comment (pllim) [reply docs/imviz/plugins.rst:104]: If you know of a way to draw lines without using markers, it would be nice but I don't think it can fit into this sprint because it is out of scope.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:32]: Good idea! I'll try that out tomorrow, along with some of your other suggestions. Thanks!

File: docs/imviz/plugins.rst
Comment (kecnry) [reply docs/imviz/plugins.rst:104]: ok, I have an idea how to do it fairly quickly but probably won't have time left this week either.  I agree its not blocking for this PR, but do think it would make this work more usable so would like to see it prioritized soon since it shouldn't be a ton of effort.

[:cat:](https://jira.stsci.edu/browse/JDAT-2324)

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (larrybradley) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:145]: For `RectanglePixelRegion` the circular annulus will intersect the corners, which is not desired (you don't want to measure background from same pixels within the aperture).  For rectangles you want the minimum radius to be to the rectangle corner:

```python
0.5 * np.sqrt(self._selected_subset.width**2 + self._selected_subset.height**2)
```

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:145]: Whooops... good catch. Thanks!

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:52]: > this means outer_radius = inner_radius + width? 

Yes. I stared at this so long, I started to assume everyone knows by heart, so thanks for your fresh pair of eyes. I'll explain.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:181]: I'll just not emit any snackbar message here for now. In the future, if people want it back for certain things, we can add it back in; that is easier than guessing what they want. I am no Mel Gibson (obscure movie reference).

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:39]:
```suggestion
            v-model.number="bg_annulus_inner_r"

```

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.vue:50]:
```suggestion
            v-model.number="bg_annulus_width"

```

**Review by pllim — COMMENTED** [review]
Use model.number

**Comment by pllim** [comment]
Oh no... wrong button... 😆

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: Now with your suggested changes, when I enter string, I get a traceback under the viewer that says:

`TraitError: The 'bg_annulus_inner_r' trait of a SimpleAperturePhotometry instance expected a float, not the str 'uyh'.`

Is this expected behavior or did I miss something?

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [inline jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:151]: I suppose an error is helpful here so that we add support for it, but does this then prevent manually entering the annulus radius/width (especially since the exception resets the subset... although that might leave the UI in an unrecoverable state)? Maybe this is a case for a snackbar and then just default to an inner radius of some fixed number.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: I'm not sure this is better as now it just silently ignores and sets some values without any feedback.  Its very easy to get the UI in a state where you're not sure what is set automatically, what was set manually, and what was set but failed to propagate.  If you can put together a list of these scenarios that you expect to trip the backend, I can look into raising them in the input validation so the user gets feedback on the element itself and without the queued snackbars.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:151]: > does this then prevent manually entering the annulus radius/width

Actually, I am not sure. I have not tested with selecting Aperture from a weird Subset that is like a result of subtracting/etc other Subsets (because mainly I don't know how to do it properly and I don't see a real science case to select it for aperture photometry).

Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:151]: If you manually load weird subsets via the Astrowidgets API, the logic I put in should make those "invisible" to the plugin in the first place, so those are not a problem. (Unless somehow my logic was overwritten since.)

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: > If you can put together a list of these scenarios

Can we defer this as a follow-up ticket? I don't think I can give you a thorough list in my rush today.

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: this should be fixed with the `type='number'` added in [3bd3a31](https://github.com/spacetelescope/jdaviz/pull/1224/commits/3bd3a3152e8b5120ebea7f513c1f1721f5767136)

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: Re: a list of these scenarios -- Okay, so coming back to this after a month... let's refresh our memories, this exception catching handles three different operations:

```python
inner_r = float(self.bg_annulus_inner_r)
reg = CircleAnnulusPixelRegion(
    self._selected_subset.center, inner_radius=inner_r,
    outer_radius=inner_r + float(self.bg_annulus_width))
self.background_value = self._calc_bg_subset_median(reg)
```

Let's break this down below.

1. `inner_r = float(self.bg_annulus_inner_r)`

This may break if `self.bg_annulus_inner_r` has string that cannot be converted to float. Maybe your JS-side check is preventing this happening from the GUI, but it is still possible from the Python side (e.g., from tests or direct call from other parts in Jdaviz), unless your template mixin thingy also prevents this?

2. `reg = CircleAnnulusPixelRegion(...)`

For this line, two things:

a. `float(self.bg_annulus_width)` can fail in the same way as `float(self.bg_annulus_inner_r)`.
b. Any checks that would fail in [CircleAnnulusPixelRegion constructor](https://github.com/astropy/regions/blob/9bd81c2d08eae589a3931f688298060d4505274d/regions/shapes/annulus.py#L83) would trigger a failure here. So, if `outer radius <= inner radius` (it is possible if you force width to be zero or negative) and less obviously, the class attribute checks that are baked into the inherited metaclasses.

3. `self.background_value = self._calc_bg_subset_median(reg)`

For this lines, several things:

a. It extracts the region mask from selected data. Region mask is defined w.r.t. *reference data*, which might not be the selected data. So if you loaded some crazy stuff, there is a chance this might fail.
b. Then it extracts the selected data values from the mask. If the above step succeeds, this might not pose any problem.
c. Finally, it calls `np.nanmedian`, which I think at most emits warning if it sees all NaNs or empty array. Not sure if we can easily crash this but I don't wanna jinx it.

### Now what?

With the current code at the time of posting this, I removed the snackbar message. So if any of the above results in error, you see background set to zero but no message on what went wrong.

And maybe this is okay, given error only happens if something is very wrong?

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (pllim) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:154]: Update: So, even after loading weird shapes using Astrowidgets API, the plugin is "smart" enough to exclude them from the dropdown (because of the "Subset" prefix check) but this might accidentally break when we implement the feature to allow users to rename subset, so I still prefer to leave this logic in.

> does this then prevent manually entering the annulus radius/width (especially since the exception resets the subset... although that might leave the UI in an unrecoverable state)?

I am not sure if this is possible currently. The `else` is just there for completeness right now. If we ever see an exception here, it is a serious bug and we need to add new code to prevent it properly. In that case, I think it is clearer to see ugly traceback because Snackbar can be easily dismissed.

File: jdaviz/utils.py
Comment (pllim) [reply jdaviz/utils.py:122]: I am not excited and code-diving bqplot again but you have a good point and you caught me being lazy. I will see what I can do here.

Comment (pllim) [reply jdaviz/utils.py]: Okay, implemented.

File: jdaviz/utils.py
Comment (pllim) [inline jdaviz/utils.py:119]: Maybe @maartenbreddels know of a better way but I couldn't find it in `bqplot` doc, example, or quick code-dive.

**Comment by pllim** [comment]
@kecnry , I believe I have addressed your comments thus far. Please let me know what you think. Thanks!

File: jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py]: I think this is a reasonable solution for now, thanks.  The most common case that you can easily trigger from the UI is a non-positive inner-radius or width.  These now raise the validation warning on the input to blame, display the background as zero, but allow the rest to be computed (all without queuing endless snackbars).  The other scenarios you brought up would likely require intentionally trying to break the plugin, and since the UI still provides feedback of what value is being adopted for the background (0), I don't think they're worth a ton of effort to manually catch now.

File: docs/imviz/plugins.rst
Comment (kecnry) [inline docs/imviz/plugins.rst:106]:
```suggestion
    and select that region using the :guilabel:`Background` drop-down menu.
```

(I think this label was renamed when creating the subset dropdowns)

**Review by kecnry — APPROVED** [review]
Just a few small suggestions for the docs.  Feature seems to work well, so let's get this in before the conflicts get messy!  

I did notice that changes to selected subsets aren't observed (background value doesn't change if the user changes the subset selected for the background), but that is out of scope here and something we need to address consistently app-wide at some point.  The adopted background value is visible to the user, so there at least won't be any surprises and they should figure out their changes weren't immediately adopted.

**Comment by pllim** [comment]
> I did notice that changes to selected subsets aren't observed (background value doesn't change if the user changes the subset selected for the background)

I don't think I can reproduce this. Are you sure you are not encountering the issue where your two subsets actually have the same median background? Can you please give me the exact workflow to reproduce this?

Anyway, if you don't think this is in scope anyway, please merge. Thanks!

**Comment by pllim** [comment]
Oh wait, we still need a second review, right?

**Comment by pllim** [comment]
So... I have decided to count @camipacifici comment on Slack from like 1-2 weeks ago as second approval. I need to get this in so I can work on the next thing. Thanks!

**Pull request merged**
